test/propagated_params_test.go

```diff
@@ -0,0 +1,161 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/test/parse"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	knativetest "knative.dev/pkg/test"
+)
+
+func TestPropagatedParams(t *testing.T) {
+	t.Parallel()
+	type tests struct {
+		name            string
+		pipelineName    string
+		taskName        string
+		pipelineRunFunc func(*testing.T, string, string, string) *v1beta1.PipelineRun
+	}
+
+	tds := []tests{{
+		name:            "propagated parameters fully",
+		pipelineName:    "propagated-parameters-fully",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamPipelineRun,
+	}, {
+		name:            "propagated parameters with inner scope",
+		pipelineName:    "propagated-parameters-inner-scope",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamWithScopePipelineRun,
+	}, {
+		name:            "propagated parameters with default inner scope",
+		pipelineName:    "propagated-parameters-default-inner-scope",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamWithDefaultScopePipelineRun,
+	}}
+
+	for _, td := range tds {
+		td := td
+		t.Run(td.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			propagatedParamFlags := requireAllGates(map[string]string{
+				"enable-api-fields": "alpha",
+			})
+			c, namespace := setup(ctx, t, propagatedParamFlags)
+
+			knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+			defer tearDown(ctx, t, c, namespace)
+
+			t.Logf("Setting up test resources for %q test in namespace %s", td.name, namespace)
+			pipelineRun := td.pipelineRunFunc(t, namespace, td.pipelineName, td.taskName)
+			prName := pipelineRun.Name
+			_, err := c.PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Failed to create PipelineRun `%s`: %s", prName, err)
+			}
+
+			t.Logf("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+			if err := WaitForPipelineRunState(ctx, c, prName, timeout, PipelineRunSucceed(prName), "PipelineRunSuccess"); err != nil {
+				t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+			}
+			t.Log("Checking if parameter replacements have been updated in the spec.")
+			cl, _ := c.PipelineRunClient.Get(ctx, prName, metav1.GetOptions{})
+			if cl.Status.PipelineSpec.Tasks[0].TaskSpec.Steps[0].Script != "echo Hello World!" {
+				t.Fatalf(`Expected replaced parameter value %s but found %s`, "echo Hello World!", cl.Status.PipelineSpec.Tasks[0].TaskSpec.Steps[0].Script)
```

Comment:
could we check the resolved pipelinerun vs expected pipelinerun instead of the script only? this will help us ensure that the spec was not modified (while ignoring the fields that can't be pre-determined like start time)